### PR TITLE
refactor(common): generalize `LookupEntryState` to `EstimatedHashSet`

### PR DESCRIPTION
**Pull request by stdrc**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Seems previously `LookupEntryState` is just a simple `EstimateSize` wrapper for `HashSet`, this PR generalize it to `EstimatedHashSet` so that other modules can reuse it in the future.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by stdrc** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15843?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#15843** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15843?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#15842** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15842?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#15841** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15841?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#15840** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15840?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @stdrc and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/common/estimate_size/src/collections/hashset.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+use std::collections::HashSet;
+use std::hash::Hash;
+
+use super::EstimatedVec;
+use crate::{EstimateSize, KvSize};
+
+#[derive(Default)]
+pub struct EstimatedHashSet<T: EstimateSize> {
+    inner: HashSet<T>,
+    kv_heap_size: KvSize,
```
Comment (BugenZhao) [inline src/common/estimate_size/src/collections/hashset.rs:24]: cc @yuhao-su May I ask what "kv" in the name is for? 😄

File: src/stream/src/executor/lookup/cache.rs [hunk 2]
```diff
@@ -45,12 +44,17 @@ impl LookupCache {
             let key = row.project(arrange_join_keys).into_owned_row();
             if let Some(mut values) = self.data.get_mut(&key) {
                 // the item is in cache, update it
+                let row = row.into_owned_row();
                 match op {
                     Op::Insert | Op::UpdateInsert => {
-                        values.insert(row.into_owned_row());
+                        if !values.insert(row) {
+                            panic!("inserting a duplicated value");
```
Comment (BugenZhao) [inline src/stream/src/executor/lookup/cache.rs:51]: It appears that we can now safely replace `panic` with `inconsistency_panic` after the refactoring.

**Review by BugenZhao — APPROVED** [review]
Looks good

File: src/stream/src/executor/lookup/cache.rs (hunk 2 above)
Comment (stdrc) [reply src/stream/src/executor/lookup/cache.rs:51]: Yes, exactly.

File: src/common/estimate_size/src/collections/hashset.rs (hunk 1 above)
Comment (stdrc) [reply src/common/estimate_size/src/collections/hashset.rs]: Just renamed the `kv_heap_size` field to `heap_size`. Maybe we should revisit the name of `KvSize` struct in the future, but let's keep it unchanged in this PR.

**Pull request merged**